tools/webhook-apicoverage/resourcetree/test_util.go

```diff
@@ -0,0 +1,182 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resourcetree
+
+//test_util contains types defined and used by types and their corresponding verification methods.
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+)
+
+const (
+	basicTypeName = "BasicType"
+	ptrTypeName = "PtrType"
+	arrayTypeName = "ArrayType"
+	otherTypeName = "OtherType"
+	combinedTypeName = "CombinedType"
+)
+
+type baseType struct {
+	field1 string
+	field2 int16
+}
+
+type ptrType struct {
+	structPtr *baseType
+	basePtr *float32
+}
+
+type arrayType struct {
+	structArr []baseType
+	baseArr []bool
+}
+
+type otherType struct {
+	structMap map[string]baseType
+	baseMap map[string]string
+}
+
+type combinedNodeType struct {
+	b baseType
+	a arrayType
+	p ptrType
+}
+
+func verifyBaseTypeNode(logPrefix string, data nodeData) error {
+	if len(data.children) != 2 {
+		return fmt.Errorf("%s Expected 2 children got only : %d", logPrefix, len(data.children))
+	}
+
+	if value, ok := data.children["field1"]; ok {
+		n := value.getData()
+		if !n.leafNode || n.fieldType.Kind() != reflect.String || n.fieldType.PkgPath() != "" || len(n.children) != 0 {
+			return fmt.Errorf("%s Unexpected field:field1 : - %v of kind : %s", logPrefix, n, n.fieldType.Kind())
```

Comment:
This error message looks confusing. I suggest  clearly logging what's the expectation and what's the actual value. Here and everywhere.